src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/TypeConstraintVisitor.cs

```diff
@@ -0,0 +1,138 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections;
+using System.Collections.Generic;
+using EnsureThat;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+using Microsoft.Health.Fhir.SqlServer.Features.Storage;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// Looks for _type parameters to determine the set of ResourceTypeIds allowed in an expression.
+    /// Assumes that <see cref="ResourceColumnPredicatePushdownRewriter"/> has not run yet on the expression.
+    /// </summary>
+    internal class TypeConstraintVisitor : DefaultSqlExpressionVisitor<(BitArray allowedTypes, ISqlServerFhirModel model), short?>
+    {
+        internal const short NoTypes = -1;
```

Comment:
>NoTypes [](http://example.com/codeflow?start=29&length=7)

nit: 

It feels a bit strange what it's not used outside of this class (tests of this class is exception).

I mean al this work to indicate what there is no types, and we will ignore that and use allAllowTypes to find what is supported types.

Reply:
By returning `NoTypes`, we are not returning null, where null means many types. When we return NoTypes, we don't touch `allowedTypes`.

Reply:
My bad, you right.
Still a bit weird to return type id and use it to only as indication of using BitArray or not.

Considering what BitArray is class we can just return null for bit array instead of returning null and bitarray.

Reply:
Done.